bulk_update_documents crashes with DoesNotExist in build_document_node when a correspondent/document type is deleted concurrently (LLM index enabled)

### Description

When the LLM index (AI features) is enabled, `documents.tasks.bulk_update_documents` can crash with `Correspondent matching query does not exist` (or `DocumentType matching query does not exist`) if the referenced correspondent or document type is deleted while the task's `update_llm_index` pass is running.

The metadata change itself is committed correctly (the foreign key is set to `NULL` as expected, no orphaned rows remain), but the LLM-index update for the affected documents is aborted, leaving their vector nodes stale until the next `document_llmindex` run. It is a small unguarded foreign-key access in `build_document_node`, present in both the latest release (3.0.3) and `main`.

### Background & real-world use case (non-technical)

Paperless-ngx can build an *LLM index* — a semantic/vector index over your documents that powers the AI features (similar-document suggestions, RAG-based Q&A). Whenever a document changes, its entry in this index is refreshed so the AI stays in sync with the library.

A very common maintenance activity is **tidying up the taxonomy** — e.g. merging duplicate correspondents (say, "ACME Ltd" and "ACME Limited" into one) or removing a document type you no longer use. When this is done across many documents at once, Paperless queues background jobs that refresh each affected document, including its LLM-index entry.

The problem: if a correspondent or document type is deleted at roughly the same time these background refresh jobs run for documents that still reference it, the LLM-index refresh **crashes** for those documents. The documents themselves are fine — the correspondent/type is removed cleanly and nothing is lost — but the AI index for those few documents is left slightly out of date until the next index run, and the logs fill with an error traceback.

In short: this hits users who **(a)** have the AI/LLM index enabled and **(b)** do bulk taxonomy cleanup. It surfaced here during a large re-classification run that reassigned/merged correspondents and document types across the whole library.

### Environment

- **Paperless-ngx:** 3.0.3 (also reproducible against `main` by code inspection — see Notes)
- **Deployment:** Docker (official `ghcr.io` image)
- **AI features:** LLM index enabled (embedding backend configured)
- **Database:** PostgreSQL

### Traceback (abridged)

```
[ERROR] [celery.app.trace] Task documents.tasks.bulk_update_documents[…] raised unexpected:
    DoesNotExist('Correspondent matching query does not exist.')
  ...
  File "src/documents/tasks.py", line 332, in bulk_update_documents
    update_llm_index(rebuild=False)
  File "src/paperless_ai/indexing.py", line 378, in update_llm_index
    nodes = build_document_node(document, chunk_size=chunk_size)
  File "src/paperless_ai/indexing.py", line 184, in build_document_node
    if document.correspondent
  File ".../django/db/models/fields/related_descriptors.py", line 224, in get_object
    ...
django.db.models.fields.related_descriptors.RelatedObjectDoesNotExist:
    Correspondent matching query does not exist.
```

The same crash occurs for `document_type` (indexing.py:187).

### Root cause

`build_document_node` accesses the related objects with an unguarded lazy load:

```python
"correspondent": document.correspondent.name if document.correspondent else None,
"document_type": document.document_type.name if document.document_type else None,
"storage_path": document.storage_path.name if document.storage_path else None,
```

`if document.correspondent` triggers a lazy DB fetch by `correspondent_id`. If that correspondent was deleted between the time the document was read for the index pass and the time `build_document_node` resolves the relation, the fetch raises `RelatedObjectDoesNotExist` instead of returning `None`, and the whole `update_llm_index` pass then aborts for the remaining documents in that task.

### Steps to reproduce

1. Enable the LLM index.
2. Assign a correspondent (or document type) to one or more documents.
3. Trigger a `bulk_edit` on those documents (which enqueues `bulk_update_documents`) and delete that correspondent/document type at around the same time.
4. The `bulk_update_documents` task raises `DoesNotExist` in `build_document_node`.

### Impact

- **No data loss:** the bulk metadata change is committed, FKs are nulled correctly, no orphaned rows.
- The LLM index nodes for the affected documents are **not updated** until the next `document_llmindex update`/`rebuild`.
- Noisy `ERROR` tracebacks during otherwise-normal bulk taxonomy maintenance.

### Proposed fix

Either of:

1. **Re-fetch the document freshly at task start** so the instance reflects the committed `NULL` FK, or
2. **Guard the relation access** in `build_document_node`, treating a missing related object the same as `None`:

```python
from django.core.exceptions import ObjectDoesNotExist

def _safe_related_name(document, field):
    try:
        rel = getattr(document, field)
    except ObjectDoesNotExist:
        return None
    return rel.name if rel else None
```

Option 2 is minimal and keeps node building resilient regardless of concurrency.

### Notes

- `main` still has the unguarded ternary access in `build_document_node`, so upgrading does not resolve this.
- Related prior work: #12899 (index locking during bulk update) and #12893 (AI indexing fixes) — the FK-access guard for `build_document_node` appears not to be covered by either.

Full raw tracebacks for both variants (`DocumentType` and `Correspondent`), from paperless-ngx 3.0.3. Task UUIDs are the only identifiers; no document content is involved.

**1) `DocumentType` variant**

```
[2026-07-26 20:30:24,715] [ERROR] [celery.app.trace] Task documents.tasks.bulk_update_documents[bc03fed8-…] raised unexpected: DoesNotExist('DocumentType matching query does not exist.')
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py", line 243, in __get__
    rel_obj = self.field.get_cached_value(instance)
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/mixins.py", line 37, in get_cached_value
    return instance._state.fields_cache[self.cache_name]
KeyError: 'document_type'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/celery/app/trace.py", line 585, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.12/site-packages/celery/app/trace.py", line 858, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/src/paperless/src/documents/tasks.py", line 332, in bulk_update_documents
    update_llm_index(
  File "/usr/src/paperless/src/paperless_ai/indexing.py", line 367, in update_llm_index
    nodes = build_document_node(document, chunk_size=chunk_size)
  File "/usr/src/paperless/src/paperless_ai/indexing.py", line 187, in build_document_node
    if document.document_type
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py", line 261, in __get__
    rel_obj = self.get_object(instance)
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py", line 224, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 635, in get
    raise self.model.DoesNotExist(
documents.models.DocumentType.DoesNotExist: DocumentType matching query does not exist.
```

**2) `Correspondent` variant**

```
[2026-07-26 20:34:37,480] [ERROR] [celery.app.trace] Task documents.tasks.bulk_update_documents[a9d5bd3c-…] raised unexpected: DoesNotExist('Correspondent matching query does not exist.')
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py", line 243, in __get__
    rel_obj = self.field.get_cached_value(instance)
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/mixins.py", line 37, in get_cached_value
    return instance._state.fields_cache[self.cache_name]
KeyError: 'correspondent'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/celery/app/trace.py", line 585, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.12/site-packages/celery/app/trace.py", line 858, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/src/paperless/src/documents/tasks.py", line 332, in bulk_update_documents
    update_llm_index(
  File "/usr/src/paperless/src/paperless_ai/indexing.py", line 378, in update_llm_index
    nodes = build_document_node(document, chunk_size=chunk_size)
  File "/usr/src/paperless/src/paperless_ai/indexing.py", line 184, in build_document_node
    if document.correspondent
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py", line 261, in __get__
    rel_obj = self.get_object(instance)
  File "/usr/local/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py", line 224, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 635, in get
    raise self.model.DoesNotExist(
documents.models.Correspondent.DoesNotExist: Correspondent matching query does not exist.
```

Note the `KeyError` → `get_object` → `DoesNotExist` chain: the FK id is set on the instance but the referenced row is gone by the time `build_document_node` resolves the relation. The DB ends up consistent (the FK is nulled via `SET_NULL`); only this in-flight resolution fails.

Side observation from the same run (context, not part of the bug): individual `bulk_update_documents` tasks took ~935–945 s each, because every task triggers a full `update_llm_index(rebuild=False)` pass over the whole library. A large bulk taxonomy edit therefore enqueues many such tasks, each doing a full-library index pass — so this crash tends to recur across a batch rather than once.

--- a/src/paperless_ai/indexing.py
+++ b/src/paperless_ai/indexing.py
@@ -5,6 +5,7 @@ from datetime import timedelta
 from typing import TYPE_CHECKING
 
 from django.conf import settings
+from django.core.exceptions import ObjectDoesNotExist
 from django.utils import timezone
 from filelock import FileLock
 from filelock import ReadWriteLock
@@ -167,6 +168,19 @@ def write_store(embed_model_name: str | None = None):
         yield store
 
 
+def _safe_related_name(document: Document, field: str) -> str | None:
+    """
+    Returns the ``name`` of a related object (correspondent, document_type,
+    storage_path), or None if the FK is unset or points at a row that has
+    since been deleted (e.g. concurrently with this call).
+    """
+    try:
+        related = getattr(document, field)
+    except ObjectDoesNotExist:
+        return None
+    return related.name if related else None
+
+
 def build_document_node(
     document: Document,
     *,
@@ -180,14 +194,10 @@ def build_document_node(
         "document_id": str(document.id),
         "title": document.title,
         "tags": [t.name for t in document.tags.all()],
-        "correspondent": document.correspondent.name
-        if document.correspondent
-        else None,
-        "document_type": document.document_type.name
-        if document.document_type
-        else None,
+        "correspondent": _safe_related_name(document, "correspondent"),
+        "document_type": _safe_related_name(document, "document_type"),
         "filename": document.filename,
-        "storage_path": document.storage_path.name if document.storage_path else None,
+        "storage_path": _safe_related_name(document, "storage_path"),
         "archive_serial_number": document.archive_serial_number,
         "created": document.created.isoformat() if document.created else None,
         "added": document.added.isoformat() if document.added else None,

--- a/src/paperless_ai/tests/test_ai_indexing.py
+++ b/src/paperless_ai/tests/test_ai_indexing.py
@@ -8,7 +8,9 @@ from django.test import override_settings
 from django.utils import timezone
 from llama_index.core.schema import MetadataMode
 
+from documents.models import Correspondent
 from documents.models import Document
+from documents.models import DocumentType
 from documents.models import PaperlessTask
 from documents.signals import document_consumption_finished
 from documents.signals import document_updated
@@ -93,6 +95,38 @@ def test_build_document_node_structured_fields_in_metadata(
         assert "created" in node.metadata
         assert "added" in node.metadata
         assert "modified" in node.metadata
+
+
+@pytest.mark.django_db
+def test_build_document_node_survives_concurrently_deleted_correspondent(
+    real_document: Document,
+) -> None:
+    """Regression test for #13314.
+
+    If a document's correspondent (or document type) is deleted after the
+    in-memory Document instance was loaded but before build_document_node
+    resolves the relation, accessing the FK must not raise -- it should
+    behave like an unset FK and produce None in the metadata instead of
+    aborting the whole indexing pass.
+    """
+    correspondent = Correspondent.objects.create(name="Stale Correspondent")
+    document_type = DocumentType.objects.create(name="Stale Type")
+    real_document.correspondent = correspondent
+    real_document.document_type = document_type
+    real_document.save()
+
+    # Re-fetch to get an instance whose correspondent/document_type relations
+    # are unresolved (not yet cached), mirroring a task that loaded the
+    # document before the concurrent deletion below.
+    stale_document = Document.objects.get(pk=real_document.pk)
+
+    correspondent.delete()
+    document_type.delete()
+
+    nodes = indexing.build_document_node(stale_document)
+    assert len(nodes) > 0
+    assert nodes[0].metadata["correspondent"] is None
+    assert nodes[0].metadata["document_type"] is None
 
 
 @pytest.mark.django_db

